### PR TITLE
140033 :  Accessing config files and storing in object

### PR DESCRIPTION
**Pull request by abhijeet14992**
This PR consist changes of  [PBI - 140033](https://dev.azure.com/ukhydro/Exchange%20Set%20Service/_workitems/edit/140033) as follows:

- Code changes to access config files and storing in object
- Added unit test cases.
- Added storage account and container

### CODE REVIEW

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService.UnitTests/Services/ConfigurationServiceTests.cs
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService.UnitTests/Services/ConfigurationServiceTests.cs:1]: EventIds in logs are missing, please add where missing

e.g. 
A.CallTo(fakeLogger).Where(call =>
                  call.Method.Name == "Log"
                  && call.GetArgument<LogLevel>(0) == LogLevel.Information
&& call.GetArgument<EventId>(1) == EventIds.BessConfigsProcessingStarted.ToEventId()
                  && call.GetArgument<IEnumerable<KeyValuePair<string, object>>>(2).ToDictionary(c => c.Key, c => c.Value)["{OriginalFormat}"].ToString() == "Bess configs processing started, Total configs count:{count}  | _X-Correlation-ID : {CorrelationId}"
                  ).MustHaveHappenedOnceExactly();

**Review by Nadeem14456 — CHANGES_REQUESTED** [review]

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService.UnitTests/Services/ConfigurationServiceTests.cs [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        {
+            fakeAzureBlobStorageClient = A.Fake<IAzureBlobStorageClient>();
+            fakeLogger = A.Fake<ILogger<ConfigurationService.Services.ConfigurationService>>();
+
+            configurationService = new ConfigurationService.Services.ConfigurationService(fakeAzureBlobStorageClient, fakeLogger);
+            dictionary = new Dictionary<string, string>();
+        }
+
+        [Test]
+        public void Does_Constructor_Throws_ArgumentNullException_When_Parameter_Is_Null()
```
[the remark this answers is not in the document]
Comment (abhijeet14992) [reply UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService.UnitTests/Services/ConfigurationServiceTests.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService.UnitTests/Services/ConfigurationServiceTests.cs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+            nullAzureBlobStorageClient.Should().ThrowExactly<ArgumentNullException>().And.ParamName.Should().Be("azureBlobStorageClient");
+
+            Action nullConfigurationServiceLogger = () => new ConfigurationService.Services.ConfigurationService(fakeAzureBlobStorageClient, null);
+
+            nullConfigurationServiceLogger.Should().ThrowExactly<ArgumentNullException>().And.ParamName.Should().Be("logger");
+        }
+
+        [Test]
+        public void WhenInvalidConfigFileIsFound_ThenDataIsNotAddedInModelList()
```
[the remark this answers is not in the document]
Comment (abhijeet14992) [reply UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService.UnitTests/Services/ConfigurationServiceTests.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService.UnitTests/Services/ConfigurationServiceTests.cs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+            A.CallTo(fakeLogger).Where(call =>
+                  call.Method.Name == "Log"
+                  && call.GetArgument<LogLevel>(0) == LogLevel.Information
+                  && call.GetArgument<IEnumerable<KeyValuePair<string, object>>>(2).ToDictionary(c => c.Key, c => c.Value)["{OriginalFormat}"].ToString() == "Bess configs processing completed | _X-Correlation-ID : {CorrelationId}"
+                  ).MustHaveHappenedOnceExactly();
+        }
+
+        [Test]
+        public void WhenValidConfigFileIsFound_ThenDataIsAddedInModelList()
```
[the remark this answers is not in the document]
Comment (abhijeet14992) [reply UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService.UnitTests/Services/ConfigurationServiceTests.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/BessConfigurationServiceJob.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    [ExcludeFromCodeCoverage]
+    public class BessConfigurationServiceJob
+    {
+        private readonly ILogger<BessConfigurationServiceJob> logger;
+        private readonly IConfigurationService configurationService;
+
+        public BessConfigurationServiceJob(ILogger<BessConfigurationServiceJob> logger, IConfigurationService configurationService)
+        {
+            this.logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            this.configurationService = configurationService;
```
[the remark this answers is not in the document]
Comment (abhijeet14992) [reply UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/BessConfigurationServiceJob.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/appsettings.json [hunk 5]
```diff
@@ -28,5 +28,9 @@
         "Service": "BESS Configuration Service",
         "NodeName": "Azure",
         "ConnectionString": ""
+    },
+    "BessStorageConfiguration": {
+        "ConnectionString": "",
+        "ContainerName": "bess-configs"
```
[the remark this answers is not in the document]
Comment (abhijeet14992) [reply UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/appsettings.json:34]: Resolved

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/AzureBlobStorageClient.cs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+                    if (blobItem.Name.EndsWith(".json", StringComparison.OrdinalIgnoreCase))
+                    {
+                        BlobClient blobClient = GetBlobClient(blobContainerClient, blobItem.Name);
+                        configs.Add(blobItem.Name, DownloadBlobContent(blobClient));
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                logger.LogError(EventIds.BessErrorOccuredWhileDownloadingConfigFromAzureStorage.ToEventId(), "Exception occured while dowloading configs from azure storage with Exception Message : {Message} | StackTrace : {StackTrace} | _X-Correlation-ID : {CorrelationId}", ex.Message, ex.StackTrace, CommonHelper.CorrelationID);
```
[the remark this answers is not in the document]
Comment (abhijeet14992) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/AzureBlobStorageClient.cs]: Log updated

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs [hunk 7]
```diff
@@ -401,13 +401,41 @@ public enum EventIds
...
+        /// </summary>
+        BessConfigParsingError = 809105,
+        /// <summary>
+        /// 809106 - Bess configs not found
+        /// </summary>
+        BessConfigsNotFound = 809106,
+        /// <summary>
+        /// 809107 - Bess error occured while downloading configs from azure storage
+        /// </summary>
+        BessErrorOccuredWhileDownloadingConfigFromAzureStorage = 809107
```
[the remark this answers is not in the document]
Comment (abhijeet14992) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService.UnitTests/Services/ConfigurationServiceTests.cs
Comment (abhijeet14992) [reply UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService.UnitTests/Services/ConfigurationServiceTests.cs:1]: EventIds in logs added

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Models/Bess/BessConfig.cs [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+        public IEnumerable<Tag> Tags { get; set; }
+        public string ReadMeSearchFilter { get; set; }
+        public int BatchExpiryInDays { get; set; }
+        public bool? IsEnabled { get; set; }
+    }
+
+    public class Tag
+    {
+        public string key { get; set; }
+        public string value { get; set; }
```
[the remark this answers is not in the document]
Comment (abhijeet14992) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Models/Bess/BessConfig.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs [hunk 9]
```diff
@@ -401,13 +401,41 @@ public enum EventIds
...
+        /// <summary>
+        /// 809105 - Bess config parsing error
+        /// </summary>
+        BessConfigParsingError = 809105,
+        /// <summary>
+        /// 809106 - Bess configs not found
+        /// </summary>
+        BessConfigsNotFound = 809106,
+        /// <summary>
+        /// 809107 - Bess error occurred while downloading configs from azure storage
```
[the remark this answers is not in the document]
Comment (abhijeet14992) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Logging/EventIds.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/ConfigurationService.cs [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+        public void ProcessConfigs()
+        {
+            try
+            {
+                Dictionary<string, string> configs = azureBlobStorageClient.GetConfigsInContainer();
+
+                logger.LogInformation(EventIds.BessConfigsProcessingStarted.ToEventId(), "Bess configs processing started, Total configs count:{count}  | _X-Correlation-ID : {CorrelationId}", configs.Keys.Count, CommonHelper.CorrelationID);
+
+                List<BessConfig> bessConfigs = new();
```
[the remark this answers is not in the document]
Comment (abhijeet14992) [reply UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/ConfigurationService.cs]: Changed List to IList

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/ConfigurationService.cs [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+                {
+                    foreach (string fileName in configs.Keys.ToList())
+                    {
+                        string content = configs[fileName];
+
+                        bool isValidJson = IsValidJson(content, fileName);
+
+                        if (isValidJson)
+                        {
+                            IList<BessConfig> bessconfig = JsonConvert.DeserializeObject<List<BessConfig>>(content)!;
```
Comment (Nadeem14456) [inline UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/ConfigurationService.cs:41]:
```suggestion
                            IList<BessConfig> bessConfig = JsonConvert.DeserializeObject<List<BessConfig>>(content)!;
```

**Review by Nadeem14456 — CHANGES_REQUESTED** [review]

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/ConfigurationService.cs (hunk 11 above)
Comment (abhijeet14992) [reply UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/ConfigurationService.cs]: Resolved

**Review by Nadeem14456 — APPROVED** [review]
LGTM.

**Review by amogh14893 — APPROVED** [review]
Reviewed DevOps Changes, LGTM as long as terraform deployment jobs are passing without any issues after latest commits.

**Review by shirinbt14926 — APPROVED** [review]
LGTM

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/AzureBlobStorageClient.cs [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+
+        private BlobContainerClient GetBlobContainerClient()
+        {
+            BlobContainerClient blobContainerClient = new(bessStorageConfiguration.ConnectionString, bessStorageConfiguration.ContainerName);
+
+            bool isExist = blobContainerClient.Exists();
+
+            if (!isExist)
+            {
+                throw new Exception("Container does not exists");
```
[the remark this answers is not in the document]
Comment (harshal11869) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Helpers/AzureBlobStorageClient.cs:57]: Ok.

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/BessConfigurationServiceJob.cs [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+                configurationService.ProcessConfigs();
+
+                logger.LogInformation(EventIds.BessConfigurationServiceCompleted.ToEventId(), "Bess Configuration Service Completed | _X-Correlation-ID : {CorrelationId}", CommonHelper.CorrelationID);
+            }
+            catch (Exception ex)
+            {
+                logger.LogError(EventIds.UnhandledException.ToEventId(), "Exception occured while processing Bess Configuration Service webjob with Exception Message : {Message} | StackTrace : {StackTrace} | _X-Correlation-ID : {CorrelationId}", ex.Message, ex.StackTrace, CommonHelper.CorrelationID);
+                throw;
+            }
```
[the remark this answers is not in the document]
Comment (harshal11869) [reply UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/BessConfigurationServiceJob.cs:35]: Ok.

**Review by harshal11869 — CHANGES_REQUESTED** [review]

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/IConfigurationService.cs [hunk 14]
```diff
@@ -0,0 +1,7 @@
+﻿namespace UKHO.BESS.ConfigurationService.Services
+{
+    public interface IConfigurationService
+    {
+        void ProcessConfigs();
```
Comment (DanOwenUKHO) [inline UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/IConfigurationService.cs:5]: Would a boolean return type be useful for this method?

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Program.cs [hunk 15]
```diff
@@ -38,8 +40,8 @@ public static async Task Main()
                 ServiceProvider serviceProvider = serviceCollection.BuildServiceProvider();
                 try
                 {
-                    var bessConfigurationServiceJob = serviceProvider.GetService<BESSConfigurationServiceJob>();
-                    await bessConfigurationServiceJob.CreateBespokeExchangeSetAsync();
+                    var bessConfigurationServiceJob = serviceProvider.GetService<BessConfigurationServiceJob>();
+                    bessConfigurationServiceJob.Start();
```
Comment (DanOwenUKHO) [inline UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Program.cs:44]: Any reason we have changed the method name to start here? 'CreateBespokeExchangeSetAsync' is more descriptive in my opinion

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/BessConfigurationServiceJob.cs [hunk 16]
```diff
@@ -0,0 +1,38 @@
+﻿using System.Diagnostics.CodeAnalysis;
+using Microsoft.Extensions.Logging;
+using UKHO.BESS.ConfigurationService.Services;
+using UKHO.PeriodicOutputService.Common.Helpers;
+using UKHO.PeriodicOutputService.Common.Logging;
+
+namespace UKHO.BESS.ConfigurationService
+{
+    [ExcludeFromCodeCoverage]
+    public class BessConfigurationServiceJob
```
Comment (DanOwenUKHO) [inline UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/BessConfigurationServiceJob.cs:10]: Why has the capitalisation changed on 'Bess'? Wasn't 'BESS' the correct capitalisation as its an acronym?

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Configuration/BessStorageConfiguration.cs [hunk 17]
```diff
@@ -0,0 +1,10 @@
+﻿namespace UKHO.PeriodicOutputService.Common.Configuration
+{
+    public class BessStorageConfiguration
```
Comment (DanOwenUKHO) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Configuration/BessStorageConfiguration.cs:3]: As other comment, should the class be capitalised as 'BESSStorageConfiguration'

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Models/Bess/BessConfig.cs [hunk 18]
```diff
@@ -0,0 +1,27 @@
+﻿using Newtonsoft.Json;
+
+namespace UKHO.PeriodicOutputService.Common.Models.Bess
+{
+    [JsonObject(ItemNullValueHandling = NullValueHandling.Ignore)]
+    public class BessConfig
```
Comment (DanOwenUKHO) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Models/Bess/BessConfig.cs:6]: As other comment should class be capitalised as 'BESSConfig'

**Review by DanOwenUKHO — CHANGES_REQUESTED** [review]
A few small queries from me

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Models/Bess/BessConfig.cs [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+        public string KeyFileType { get; set; }
+        public IEnumerable<string> AllowedUsers { get; set; }
+        public IEnumerable<string> AllowedUserGroups { get; set; }
+        public IEnumerable<Tag> Tags { get; set; }
+        public string ReadMeSearchFilter { get; set; }
+        public int BatchExpiryInDays { get; set; }
+        public bool? IsEnabled { get; set; }
+    }
+
+    public class Tag
```
[the remark this answers is not in the document]
Comment (Nadeem14456) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Models/Bess/BessConfig.cs:22]: @abhijeet14992 please do the needful.

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/IConfigurationService.cs (hunk 14 above)
Comment (Nadeem14456) [reply UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/Services/IConfigurationService.cs:5]: not at the moment.

File: UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/BessConfigurationServiceJob.cs (hunk 16 above)
Comment (Nadeem14456) [reply UKHO.PeriodicOutputService/UKHO.BESS.ConfigurationService/BessConfigurationServiceJob.cs:10]: keep it consistent with existing code.

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Models/Bess/BessConfig.cs (hunk 18 above)
Comment (Nadeem14456) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Models/Bess/BessConfig.cs:6]: keep it consistent with existing code.

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Configuration/BessStorageConfiguration.cs (hunk 17 above)
Comment (Nadeem14456) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Configuration/BessStorageConfiguration.cs:3]: keep it consistent with existing code.

**Review by harshal11869 — APPROVED** [review]
Looks Good.

**Review by DanOwenUKHO — APPROVED** [review]

**Pull request merged**